GitCommands/ArgumentBuilderExtensions.cs

```diff
@@ -300,5 +301,92 @@ public static void Add(this ArgumentBuilder builder, [CanBeNull, ItemCanBeNull]
                 builder.Add(objectId);
             }
         }
+
+        /// <summary>
+        /// Split arguments exceeding max length into multiple batches.
+        /// Windows by default limit arguments length less than 32767 <see cref="short.MaxValue"/>.
+        /// Implementation by <see cref="System.Diagnostics.Process"/> will have file path included in command line arguments,
+        /// as well as added quotation and space characters, so we need base length to account for all these added characters
+        /// <see href="https://referencesource.microsoft.com/#system/services/monitoring/system/diagnosticts/Process.cs,1944"/>
+        /// </summary>
+        /// <param name="builder">Argument builder instance.</param>
+        /// <param name="arguments">Arguments.</param>
+        /// <param name="baseLength">Base executable file and command line length.</param>
+        /// <param name="maxLength">Command line max length. Default is 32767 - 1 on Windows.</param>
+        /// <returns>Array of batch arguments split by max length.</returns>
+        public static List<BatchArgumentItem> BuildBatchArguments(this ArgumentBuilder builder, IEnumerable<string> arguments, int? baseLength = null, int maxLength = short.MaxValue)
+        {
+            // 3: double quotes + ' '
+            // '"git.exe" ' is always included in final command line arguments
+            if (!baseLength.HasValue)
+            {
+                baseLength = AppSettings.GitCommand.Length + 3;
+            }
+
+            var baseCommand = builder.ToString();
+            if (baseLength + baseCommand.Length >= maxLength)
+            {
+                throw new ArgumentException($"Git base command \"{baseCommand}\" always reached max length of {maxLength} characters.", nameof(baseCommand));
+            }
+
+            // Clone command as argument builder
+            var tmpBuilder = new ArgumentBuilder() { baseCommand };
```

Comment:
Consider inlining `tmpBuilder`. For my taste it makes the `Aggregate` call easier to reason about.

Reply:
I'll fix this in #7218 and do a merge.

Reply:
`tmpBuilder` have been removed in #7218 to avoid extra allocation. I've merged the changes into this as well.